### ISSUE TITLE
groundTexture image format has inverted R and B channels

### ISSUE THREAD

**Issue by titos-carrasco** [enki-community/enki#74]

**Comment by titos-carrasco** [comment]
I'm testing enki viewer using the enkiplayground app and the image below (Equinoxe Infinity Album - Jean Michel Jarre) as background

![image](https://user-images.githubusercontent.com/4842986/136232411-bcd0cc59-47b5-428f-8cbb-4597a1185892.png)

But the image does not shows properly
![image](https://user-images.githubusercontent.com/4842986/136232791-6f0e8f05-b8e4-48d1-9a29-7f49b2be3408.png)

The next patch (Playground.cpp) fix the problem

```
@@ -332,11 +332,14 @@ int main(int argc, char *argv[])
 	
 	// Create the world and the viewer
 	bool igt(app.arguments().size() > 1);
 	QImage gt;
 	if (igt)
-		gt = QGLWidget::convertToGLFormat(QImage(app.arguments().last()));
+	{
+		QImage img(app.arguments().last());
+		if( img.format() ) gt = img.mirrored( false, true );
+	}
 	igt = !gt.isNull();
 	#if QT_VERSION >= QT_VERSION_CHECK(4,7,0)
 	const uint32_t *bits = (const uint32_t*)gt.constBits();
 	#else
 	uint32_t *bits = (uint32_t*)gt.bits();

```
![image](https://user-images.githubusercontent.com/4842986/136233357-b1486751-d2af-41f0-a60c-0812fce04333.png)

You can use jpg, gif, svg, etc image formats (RGB - needs the qt image format filters) - a 200cm x 200cm and 18dpi image works fine with low memory consumption

BUT:
- Viewer.cpp (line 873) uses GL_BGRA for glTexImage2D(). needs to change?. I changed GL_BGRA to GL_RGBA and image show with bad colors. why?
- PhysicalEngine.cpp (line 784) uses Color::fromARGB() in getGroundColor(). needs to change? (something like fromRGBA)
- getGroundColor() is used only in GroundSensor.cpp (line 96). needs some fix?

Saludos

**Comment by titos-carrasco** [comment]
Some test and NO need to change  nothinge else (only mirror the background)

![image](https://user-images.githubusercontent.com/4842986/136241571-a70d9c3c-ee78-43b3-84c4-111b30c35372.png)

**Comment by stephanemagnenat** [comment]
I am not sure to understand, is the problem that the background image has the red and blue channels inverted?

**Comment by stephanemagnenat** [comment]
It seems that indeed, [that line](https://github.com/enki-community/enki/blob/c011839ef33d836a166f955919c36c282069ce7a/viewer/Viewer.cpp#L873) should use `GL_RGBA` instead of `GL_BGRA` according to [QGLWidget's convertToGLFormat](https://doc.qt.io/qt-5/qglwidget.html#convertToGLFormat) documentation.

It seems that I introduced a change in [that commit](https://github.com/enki-community/enki/commit/a984bbd1c8b3d503331c15291314a98043ecfa38) and somewhat forgot about the texture being loaded from file. Probably the easiest is to fix the data on load after the call to `convertToGLFormat`, otherwise it is an API breakage of Enki and I'm sure it will break downstream users.

**Comment by titos-carrasco** [comment]
Current background image - blue: 0,0,1  - cyan: 0,1,1
![image](https://user-images.githubusercontent.com/4842986/136252899-cd3b9559-2a98-43bb-be2a-ee59d81bc881.png)

With "QGLWidget::convertToGLFormat()" when preparing groundTexture and robot over "**cyan**", gound sensors (World::getGroundColor) shows inverted values for channels
![image](https://user-images.githubusercontent.com/4842986/136255130-6cd2e76a-d073-4018-b6de-a70fc73a8baf.png)

Without "QGLWidget::convertToGLFormat()"  when preparing groundTexture (just mirroring the image) and robot over cyan, gound sensors (World::getGroundColor) shows correct values
![image](https://user-images.githubusercontent.com/4842986/136253948-5caf1e51-39ee-4096-851b-f1a7117eca66.png)

In this last case no need to change fromARGB or GL_BGRA ¿**why**?

**Comment by stephanemagnenat** [comment]
As its documentation says, `QGLWidget::convertToGLFormat()` converts to GL_RGBA but the viewer, and the robot's sensors, expect the equivalent of GL_BGRA. So the hack I propose is to still call `convertToGLFormat()`, to have a known but wrong order, than pass through the array to invert R and B.

**Comment by titos-carrasco** [comment]
Many images in QImage are stored using a 32-bit ARGB format (`QImage::Format_ARGB32` - 0xAARRGGBB). **In little endian means **BRGA****

This is why in Viewer.cpp (line 873 - initializeGL ) "groundTexture" is read as GL_UNSIGNED_BYTE  **GL_BGRA**  in glTexImage2D() - because read byte 0 first, then byte 1 and so on

In PhysicalEngine.cpp (line 783)   getGroundColor() reads ground texture data as **uint32_t** and this is why returns "Color::fromARGB" (`QImage::Format_ARGB32` format) 

So, no needs to change anything just:

1. don't use QGLWidget::convertToGLFormat() when preparing the background image (just mirror it)
2. ese QImage::Format_ARGB32 image format for any background
3. reduce dpi for lower memory consumption

Saludos

**Comment by titos-carrasco** [comment]
Just don't use `convertToGLFormat()` in:
- python/enki.cpp:195
- examples/playground/Playground.cpp:337

Saludos

**Comment by stephanemagnenat** [comment]
The reason there is the call to `convertToGLFormat` in the first place is that the loaded image can have any format, so we need to convert to the one we will use. Maybe another conversation is bether indeed, as long as we have guarantees on not having any padding between lines.

**Comment by titos-carrasco** [comment]
In that case, and considering that the GL format used in `convertToGLFormat ` is GL_RGBA:

- Viewer.cpp (line 871): format parameter in `glTexImage2D `must be GL_RGBA (to display the background image in the right way)
- PhysycalEngine.cpp (line 784): `getGroundColor `,must returns `Color::fromABGR`

Saludos

**Comment by stephanemagnenat** [comment]
I do not want to change the format of the ground data stored within Enki itself, as this would break the API and the projects using Enki. So we should convert to the current expected format on load, with 32 bpp hopefully there should be no padding, so calling `QImage::convertToFormat` with `QImage::Format_ARGB32` instead of using `convertToGLFormat` should work.

**Comment by stephanemagnenat** [comment]
Fixed by bb486bf.

**Comment by titos-carrasco** [comment]
Works fine :-)
